core/arch/arch.c

```diff
@@ -3383,17 +3383,42 @@ dr_mcontext_to_priv_mcontext(priv_mcontext_t *dst, dr_mcontext_t *src)
                 return false;
         }
         if (TEST(DR_MC_MULTIMEDIA, src->flags)) {
-            /* XXX i#1312: The structure size will double on 64-bit UNIX (possibly
-             * Windows) builds. A corresponding size check will be added with a
-             * future patch.
-             */
 #ifdef X86
             if (src->size > offsetof(dr_mcontext_t, simd)) {
-                if (src->size > offsetof(dr_mcontext_t, simd) +
-                        MCXT_NUM_SIMD_SSE_AVX_SLOTS * YMM_REG_SIZE) {
+                if (MCXT_NUM_SIMD_SLOTS > MCXT_NUM_SIMD_SSE_AVX_SLOTS &&
+                    src->size > offsetof(dr_mcontext_t, simd) +
+                            MCXT_NUM_SIMD_SSE_AVX_SLOTS * ZMM_REG_SIZE) {
+                    if (src->size < offsetof(dr_mcontext_t, simd) + sizeof(dst->simd))
+                        return false;
+                    /* Full copy, UNIX 64-bit. XXX i#1312: We don't support AVX-512
+                     * extended number of registers in 64-bit Windows yet.
+                     */
+                    memcpy(&dst->simd, &src->simd, sizeof(dst->simd));
+                } else if (MCXT_NUM_SIMD_SLOTS > MCXT_NUM_SIMD_SSE_AVX_SLOTS &&
+                           src->size > offsetof(dr_mcontext_t, simd) +
+                                   MCXT_NUM_SIMD_SSE_AVX_SLOTS * YMM_REG_SIZE) {
+                    if (src->size < offsetof(dr_mcontext_t, simd) +
+                            MCXT_NUM_SIMD_SSE_AVX_SLOTS * ZMM_REG_SIZE)
+                        return false;
+                    /* Backwards compatibility copy from old ZMM_REG_SIZE format w/o
+                     * AVX-512, UNIX 64-bit. XXX i#1312: We don't support AVX-512
+                     * extended number of registers in 64-bit Windows yet.
+                     */
+                    memcpy(&dst->simd, &src->simd,
+                           MCXT_NUM_SIMD_SSE_AVX_SLOTS * ZMM_REG_SIZE);
+                } else if (MCXT_NUM_SIMD_SLOTS == MCXT_NUM_SIMD_SSE_AVX_SLOTS &&
+                           src->size > offsetof(dr_mcontext_t, simd) +
+                                   MCXT_NUM_SIMD_SSE_AVX_SLOTS * YMM_REG_SIZE) {
+                    if (src->size < offsetof(dr_mcontext_t, simd) + sizeof(dst->simd))
+                        return false;
+                    /* Full copy if build is not UNIX 64-bit. */
```

Comment:
Again, wouldn't a full copy be the up-front == case?

Reply:
See above.

Reply:
Re-phrased/clarified.